[GTK][WPE] Add API to Damage to iterate and get the non-empty rectangles

carlosgcampos opened the pull request:
#### 4ccbfa1856505352185aa3dac2b5c154ba70aaa9
<pre>
[GTK][WPE] Add API to Damage to iterate and get the non-empty rectangles
<a href="https://bugs.webkit.org/show_bug.cgi?id=294015">https://bugs.webkit.org/show_bug.cgi?id=294015</a>

Reviewed by Adrian Perez de Castro.

This patch includes many of the changes added in 294236@main to make
Damage iterable that were reverted in 294557@main, but without making
Damage iterable again, since we need to investigate what caused the
artifacts. In the meantime we add nonEmptyRects() to get a copy of
the rects without the empty rectangles and forEachNonEmptyRect() to
iterate the rects skipping the empty ones.

Canonical link: <a href="https://commits.webkit.org/296137@main">https://commits.webkit.org/296137@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42e343974ab99769d049a623be341deee76f98b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106871 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26601 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16998 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112077 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57452 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108887 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27269 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35102 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81158 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109852 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21652 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96422 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61499 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21094 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14531 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56900 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90993 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14558 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115056 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33987 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25079 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90217 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34353 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92653 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89927 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34854 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12668 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29686 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17382 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33911 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39346 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33658 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37011 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35272 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -101,29 +101,131 @@ class Damage {
+        case Mode::Rectangles:
+            return m_rects.rects;
+        case Mode::BoundingBox:
+            return { m_minimumBoundingRectangle };
```
perhaps we could do the same as in case of `size()` and return `{ }` if mbr is empty?

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -101,29 +101,131 @@ class Damage {
+                return { m_minimumBoundingRectangle };
+            break;
+        case Mode::Full:
+            return { m_rect };
```
I think we're not validating `m_rect` in constructor, so it may be empty technically. As the same problem happens in `forEachNonEmptyRect` perhaps we could add an assert to the constructor instead of checking `isEmpty` here and in `forEachNonEmptyRect`

Scony replied:
Now that I think of it, perhaps we could split this effort into 2 PRs and in this one we could just add an assert here just to make sure that `m_rect` is not empty while we're returning it. In the case of problematic global-coordinate-space damages, as long as we're not using "full" damage we should be good.

carlosgcampos replied:
Actually, I've noticed we are checking whether m_rect is empty in several places, so for now we are handling that case. In all the modes, we return early from all add methods when the rect is empty, so Rects was already returning an empty vector in all the cases.

Scony reviewed (commented):
LGTM

carlosgcampos committed:
[GTK][WPE] Add API to Damage to iterate and get the non-empty rectangles
https://bugs.webkit.org/show_bug.cgi?id=294015

Reviewed by Adrian Perez de Castro.

This patch includes many of the changes added in 294236@main to make
Damage iterable that were reverted in 294557@main, but without making
Damage iterable again, since we need to investigate what caused the
artifacts. In the meantime we add nonEmptyRects() to get a copy of
the rects without the empty rectangles and forEachNonEmptyRect() to
iterate the rects skipping the empty ones.

Canonical link: https://commits.webkit.org/296137@main
